common/serial/port_io.h

```diff
@@ -0,0 +1,43 @@
+#ifndef PORT_IO__H
+#define PORT_IO__H
+
+#include <arch.h>
+
+class port_io {
+
+public:
+	port_io(arch_handler *arch) : m_arch(arch) 
```

Comment:
in general I would stay away from pointers. In C++, you can do subclassing without pointers just fine by passing everything by reference. The only time I would "point" to memory is when your talking to physical memory itself (i.e. an array, or to a mmio space, etc...). 

To get subclassing to work, just pass by reference. You will not be able to use pure virtual classes, so the base class will have to provide a default implementation (usually just ignore/error out), but the overrides will work just fine for the subclasses.